### ISSUE TITLE
proposal: adjust unused_local_variables and no_leading_underscores_for_local_identifiers

### ISSUE THREAD

**Issue by jamesderlin** [dart-lang/linter#3350]
In https://github.com/dart-lang/linter/issues/3126, @srawlins and I suggest that `no_leading_underscores_for_local_identifiers` *always* ignore plain underscores rather than ignoring them only in certain cases.

I think maybe it should go even further:
* Change the `no_leading_underscores_for_local_identifiers` lint to complain about all local identifiers that start with an underscore (including `_`, `__`, etc.) *if* they are used later. (Or maybe add a new but conflicting lint.)
* Change the `unused_local_variables` diagnostic to ignore all local identifiers that start with an underscore, not just `_`, `__`, etc.

The intention is:
* To allow people to indicate that a name is intentionally unused without complaint.
* To enforce that names indicated as unused actually *are* unused. (I don't think any lints currently check this?)
* To still be able to give names to unused things.  For example, even when a callback parameter is unused, IMO it sometimes can still be useful to name it so that readers can easily tell *what* parameter is intended to be unused:
    ```dart
    invokeWithCallback((_a, r, g, b) { ... });
    ```

**Comment by matanlurey** [comment]
@srawlins @pq Do you have any idea the direction you'd like to take with this request?

I'd be happy to send a patch, but don't want to go too far/not far enough if there is not consensus.

My two cents: @jamesderlin's proposal as-is SGTM. As I understand it:

> Ignore `^[_\w*]` instead of `$[_*]` for local identifiers (simplification of the rule)

**Comment by srawlins** [comment]
> Change the `no_leading_underscores_for_local_identifiers` lint to complain about all local identifiers that start with an underscore (including `_`, `__`, etc.) if they are used later. (Or maybe add a new but conflicting lint.)

We can't change the existing lint as it is there to guide towards Effective Dart style (["DON'T use a leading underscore for identifiers that aren't private."](https://dart.dev/guides/language/effective-dart/style#dont-use-a-leading-underscore-for-identifiers-that-arent-private)). But I don't see us introducing a new lint rule either.

> Change the unused_local_variables diagnostic to ignore all local identifiers that start with an underscore, not just _, __, etc.

The mechanism to ignore a diagnostic is the `// ignore:` comment.

> The intention is:

> * To allow people to indicate that a name is intentionally unused without complaint.

The `// ignore:` comment is the mechanism to indicate this.

> * To enforce that names indicated as unused actually are unused. (I don't think any lints currently check this?)

I don't think any lint rules check this. Probably a good idea.

> * To still be able to give names to unused things. For example, even when a callback parameter is unused, IMO it sometimes can still be useful to name it so that readers can easily tell what parameter is intended to be unused:
`invokeWithCallback((_a, r, g, b) { ... });`

That would be quite a departure from cur current patterns. The dart language specifies that non-local variables beginning with `_` are "private", and readers know what that underscore means. To add another convention, that local (or otherwise?) variables are "intentionally unused" would require more buy-in, I think. Maybe a language issue?
